snuba/optimize.py

```diff
@@ -292,18 +341,58 @@ def optimize_partitions(
 
         query = (query_template % args).strip()
         logger.info(f"Optimizing partition: {partition}")
-        start = time.time()
 
         # Update tracker before running clickhouse.execute since its possible
         # that the connection can get disconnected while the server is still
         # executing the optimization. In that case we don't want to
         # run optimization on the same partition twice.
         if tracker:
-            tracker.update_completed_partitions(partition)
+            tracker.update_scheduled_partitions(partition)
 
+        # if theres a merge in progress for this partition, wait for it to finish
+        while True:
+            busy_merging, estimated_sleep_time = is_busy_merging(
+                clickhouse, database, table
+            )
+            if not busy_merging:
+                break
+            else:
+                time.sleep(OPTIMIZE_BASE_SLEEP_TIME + estimated_sleep_time)
```

Comment:
I think you can be more aggressive here. If Clickhosue tells it will take 4 hours. You can still check once in a while. The elapsed time speeds up a lot if the load diminishes. Just check every 5 minutes and call it a day.